src/store.rs

```diff
@@ -132,668 +45,56 @@ impl Default for StringSet {
 
-    fun_store: IndexSet<(Ptr<F>, Ptr<F>, Ptr<F>)>,
+    pub fun_store: IndexSet<(Ptr<F>, Ptr<F>, Ptr<F>)>,
 
-    pub(crate) sym_store: StringSet,
+    pub sym_store: StringSet,
 
     // Other sparse storage format without hashing is likely more efficient
-    pub(crate) num_store: IndexSet<Num<F>>,
-
-    str_store: StringSet,
-    thunk_store: IndexSet<Thunk<F>>,
-    call0_store: IndexSet<(Ptr<F>, ContPtr<F>)>,
-    call_store: IndexSet<(Ptr<F>, Ptr<F>, ContPtr<F>)>,
-    call2_store: IndexSet<(Ptr<F>, Ptr<F>, ContPtr<F>)>,
-    tail_store: IndexSet<(Ptr<F>, ContPtr<F>)>,
-    lookup_store: IndexSet<(Ptr<F>, ContPtr<F>)>,
-    unop_store: IndexSet<(Op1, ContPtr<F>)>,
-    binop_store: IndexSet<(Op2, Ptr<F>, Ptr<F>, ContPtr<F>)>,
-    binop2_store: IndexSet<(Op2, Ptr<F>, ContPtr<F>)>,
-    if_store: IndexSet<(Ptr<F>, ContPtr<F>)>,
-    let_store: IndexSet<(Ptr<F>, Ptr<F>, Ptr<F>, ContPtr<F>)>,
-    letrec_store: IndexSet<(Ptr<F>, Ptr<F>, Ptr<F>, ContPtr<F>)>,
-    emit_store: IndexSet<ContPtr<F>>,
-
-    opaque_map: dashmap::DashMap<Ptr<F>, ScalarPtr<F>>,
+    pub num_store: IndexSet<Num<F>>,
+
+    pub str_store: StringSet,
+    pub thunk_store: IndexSet<Thunk<F>>,
+    pub call0_store: IndexSet<(Ptr<F>, ContPtr<F>)>,
+    pub call_store: IndexSet<(Ptr<F>, Ptr<F>, ContPtr<F>)>,
+    pub call2_store: IndexSet<(Ptr<F>, Ptr<F>, ContPtr<F>)>,
+    pub tail_store: IndexSet<(Ptr<F>, ContPtr<F>)>,
+    pub lookup_store: IndexSet<(Ptr<F>, ContPtr<F>)>,
+    pub unop_store: IndexSet<(Op1, ContPtr<F>)>,
+    pub binop_store: IndexSet<(Op2, Ptr<F>, Ptr<F>, ContPtr<F>)>,
+    pub binop2_store: IndexSet<(Op2, Ptr<F>, ContPtr<F>)>,
+    pub if_store: IndexSet<(Ptr<F>, ContPtr<F>)>,
+    pub let_store: IndexSet<(Ptr<F>, Ptr<F>, Ptr<F>, ContPtr<F>)>,
+    pub letrec_store: IndexSet<(Ptr<F>, Ptr<F>, Ptr<F>, ContPtr<F>)>,
+    pub emit_store: IndexSet<ContPtr<F>>,
+
+    pub opaque_ptrs: IndexSet<ScalarPtr<F>>,
+    pub opaque_cont_ptrs: IndexSet<ScalarContPtr<F>>,
+
     /// Holds a mapping of ScalarPtr -> Ptr for reverse lookups
-    pub(crate) scalar_ptr_map: dashmap::DashMap<ScalarPtr<F>, Ptr<F>, ahash::RandomState>,
+    pub scalar_ptr_map: dashmap::DashMap<ScalarPtr<F>, Ptr<F>, ahash::RandomState>,
     /// Holds a mapping of ScalarPtr -> ContPtr<F> for reverse lookups
-    scalar_ptr_cont_map: dashmap::DashMap<ScalarContPtr<F>, ContPtr<F>, ahash::RandomState>,
+    pub scalar_ptr_cont_map: dashmap::DashMap<ScalarContPtr<F>, ContPtr<F>, ahash::RandomState>,
 
     /// Caches poseidon hashes
-    poseidon_cache: PoseidonCache<F>,
+    pub poseidon_cache: PoseidonCache<F>,
     /// Contains Ptrs which have not yet been hydrated.
-    dehydrated: Vec<Ptr<F>>,
-    dehydrated_cont: Vec<ContPtr<F>>,
-    opaque_raw_ptr_count: usize,
+    pub dehydrated: Vec<Ptr<F>>,
+    pub dehydrated_cont: Vec<ContPtr<F>>,
 
-    pointer_scalar_ptr_cache: dashmap::DashMap<Ptr<F>, ScalarPtr<F>>,
+    pub pointer_scalar_ptr_cache: dashmap::DashMap<Ptr<F>, ScalarPtr<F>>,
```

Comment:
We could explore merging `pointer_scalar_ptr_cache` with `scalar_ptr_map` into a bidirectional `Ptr <-> SPtr` map